util/election.go

```
@@ -0,0 +1,112 @@
+	return func(response http.ResponseWriter, request *http.Request) {
+		switch request.Method {
+		case http.MethodGet:
+			leader, err := r.IsLeader()
+			if err != nil {
+				log.Error("msg", "Failed on leader check", "err", err)
+				http.Error(response, err.Error(), http.StatusInternalServerError)
+				return
+			}
+			fmt.Fprintf(response, "%v", leader)
+		case http.MethodPost:
+			leader, err := r.BecomeLeader()
+			if err != nil {
+				log.Error("msg", "Failed to become a leader", "err", err)
+				http.Error(response, err.Error(), http.StatusInternalServerError)
+				return
+			}
+			fmt.Fprintf(response, "%v", leader)
+		default:
+			log.Error("msg", "Request method not supported")
+			http.Error(response, "Request method not supported", http.StatusInternalServerError)
+		}
+	}
+}
+
+func (r *RestElection) handleResign() http.HandlerFunc {
+	return func(response http.ResponseWriter, request *http.Request) {
+		if request.Method == http.MethodPost {
+			err := r.Resign()
+			if err != nil {
+				log.Error("err", err)
+				http.Error(response, "Failed to resign", http.StatusInternalServerError)
+				return
+			}
+			response.WriteHeader(http.StatusOK)
+		} else {
+			log.Error("msg", "Request method not supported")
+			http.Error(response, "Request method not supported", http.StatusInternalServerError)
+		}
+
+	}
+}
+
+func (r *RestElection) Id() string {
+	return ""
+}
+
+func (r *RestElection) BecomeLeader() (bool, error) {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	if r.leader {
+		log.Debug("msg", "Instance is already a leader")
+		return r.leader, nil
+	}
+	r.leader = true
+	log.Info("msg", "Instance became a leader")
+	return r.leader, nil
+}
+
+func (r *RestElection) IsLeader() (bool, error) {
+	r.mutex.RLock()
+	defer r.mutex.RUnlock()
+	return r.leader, nil
+}
+
+func (r *RestElection) Resign() error {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	if !r.leader {
+		log.Debug("msg", "Can't resign when not a leader")
+		return nil
```

Should this path return some information to the client so it knows it's telling a non-leader to resign?

---

I've increased log level to `warn` so the one inspecting the logs should notice a problem. I am not sure if we should return an error because from our perspective the system is in proper state, however I am not strongly opinionated here...